sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,210 @@
+                return Mono.error(exception);
+            });
+        }
+
+        if (logger.isDebugEnabled()) {
+            logger.debug("cache[{}] doesn't exist, computing new value", key);
+        }
+        AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(singleValueInitFunc);
+        this.values.putIfAbsent(key, asyncLazyWithRefresh);
+        AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+        return result.getValueAsync().onErrorResume(
+            (exception) -> {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("cache[{}] resulted in error", key, exception);
+                }
+                // Remove the failed task from the dictionary so future requests can send other calls..
+                this.remove(key);
+                return Mono.error(exception);
+            }
+        );
+    }
+
+    public void set(TKey key, TValue value) {
+        if (logger.isDebugEnabled()) {
+            logger.debug("set cache[{}]={}", key, value);
+        }
+        AsyncLazyWithRefresh<TValue> updatedValue = new AsyncLazyWithRefresh<TValue>(value);
+        this.values.put(key, updatedValue);
+    }
+
+    public void remove(TKey key) {
+        values.remove(key);
+    }
+
+    /**
+     * This is AsyncLazy that has an additional Task that can
+     * be used to update the value. This allows concurrent requests
+     * to use the stale value while the refresh is occurring.
+     */
+    private class AsyncLazyWithRefresh<TValue> {
+        private final Function<TValue, Mono<TValue>> createValueFunc;
+        private final AtomicBoolean removeFromCache = new AtomicBoolean(false);
+        private AtomicReference<Mono<TValue>> value;
+        private final AtomicReference<Mono<TValue>> refreshInProgress;
+
+        public AsyncLazyWithRefresh(TValue value) {
+            this.createValueFunc = null;
+            if (this.value != null) {
+                this.value.set(Mono.just(value));
+            }
+            this.refreshInProgress = new AtomicReference<>();
+        }
+
+        public AsyncLazyWithRefresh(Function<TValue, Mono<TValue>> taskFactory) {
+            this.createValueFunc = taskFactory;
+            this.value = new AtomicReference<>();
+            this.refreshInProgress = new AtomicReference<>();
+        }
+
+        public Mono<TValue> getValueAsync() {
+            this.value.compareAndSet(null, this.createValueFunc.apply(null));
+
+            return this.value.get().cache();
+        }
+
+        public Mono<TValue> value() {
+            return value.get();
+        }
+
+        public Mono<TValue> createAndWaitForBackgroundRefreshTaskAsync(Function<TValue, Mono<TValue>> createRefreshFunction) {
+            Mono<TValue> valueMono = this.value.get();
+            AtomicReference<TValue> originalValue = new AtomicReference<>();
+
+            return valueMono.flatMap(value -> {
+                originalValue.set(value);
```

Comment:
hmm, I think it can be simplified into:
```
valueMono.flatMap(
  value -> {
        if(this.refreshInProgress.compareAndSet(null, createRefreshFunction.apply(value))) {
  }
```

Reply:
Updated the code.